torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,159 @@ namespace rpc {
+
+    // We must continue sleeping as long as the RPC Agent is running and when
+    // either the Retry Map is empty, or when the Retry Map's earliest expiring
+    // RPC is set to be retried in the future.
+    if (rpcRetryMap_.empty()) {
+      rpcRetryMapCV_.wait(lock, [&, this] {
+        return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+      });
+    }
+
+    steady_clock_time_point earliestTimeout;
 
-RpcAgent::~RpcAgent() = default;
+    if (!rpcRetryMap_.empty()) {
+      earliestTimeout = rpcRetryMap_.begin()->first;
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
+        return !(rpcAgentRunning_.load());
+      });
+    }
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+
+    // Updating these since something may have been added to the map while this
+    // thread was sleeping.
+    earliestTimeout = rpcRetryMap_.begin()->first;
+    auto& earliestRpcList = rpcRetryMap_.begin()->second;
+
+    // We iterate through all the RPC's set to be retried at the current
+    // timepoint, resend those RPC's, and add a callback that either schedules
+    // the RPC for a future retry or marks it with success/error depending on
+    // the outcome of the current send. Then, we clean up the rpcRetryMap_.
+    for (auto it = earliestRpcList.begin(); it != earliestRpcList.end();
+         /* no increment */) {
+      auto& earliestRpc = *it;
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
+      earliestRpc->retryCount_++;
+
+      fm->addCallback([this, newTime, earliestRpc](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        rpcRetryCallback(message, futErr, newTime, earliestRpc);
+      });
+
+      // We have either scheduled the RPC for a future retry, or we've marked
+      // it complete upon success, or we've marked it with an error after
+      // maxRetries unsuccessful send attempts. Thus, we can remove it from the
+      // retry Map for the current timepoint.
+      it = earliestRpcList.erase(it);
+    }
+
+    // If there are no more RPC's set to be retried at the current timepoint,
+    // we can remove the corresponsing unordered_set from the retry map.
+    if (earliestRpcList.empty()) {
+      rpcRetryMap_.erase(earliestTimeout);
+    }
+  }
+}
+
+void RpcAgent::rpcRetryCallback(
+    const rpc::Message& message,
+    const c10::optional<utils::FutureError>& futErr,
+    steady_clock_time_point newTime,
+    std::shared_ptr<RpcRetryInfo> earliestRpc) {
+  if (futErr) {
+    // Adding one since we want to include the original send as well and not
+    // just the retry count.
+    LOG(INFO) << "Send try " << std::to_string(earliestRpc->retryCount_ + 1)
+              << " failed";
+    if (earliestRpc->retryCount_ < earliestRpc->maxRetries_) {
```

Comment:
We essentially decided all non-application-layer errors are retryable, and since `sendWithRetries` will only be used for RRef internal messages, application-level errors are bugs we should fix. We're taking a closer look at this in a separate PR (#32957), and so I've removed the try-catch from process_group_agent.cpp entirely in order to keep all error-catching related changes in the other PR.